### PR TITLE
planner, util/ranger: allow filter condition pushing down to IndexScan for prefix index

### LINKED ISSUE #21145: Allow filter condition push down to IndexScan for prefix index.
## Feature Request

Allow filter condition push down to IndexScan for prefix index.

```
mysql> create table t (a text, b text, index a_b (a(255), b(255)));
Query OK, 0 rows affected (0.01 sec)

mysql> explain select * from t where a between 'a' and 'b' and b = 'b';
+-------------------------------+---------+-----------+--------------------------+---------------------------------------------------------+
| id                            | estRows | task      | access object            | operator info                                           |
+-------------------------------+---------+-----------+--------------------------+---------------------------------------------------------+
| IndexLookUp_11                | 0.25    | root      |                          |                                                         |
| ├─IndexRangeScan_8(Build)     | 250.00  | cop[tikv] | table:t, index:a_b(a, b) | range:["a","b"], keep order:false, stats:pseudo         |
| └─Selection_10(Probe)         | 0.25    | cop[tikv] |                          | eq(test.t.b, "b"), ge(test.t.a, "a"), le(test.t.a, "b") |
|   └─TableRowIDScan_9          | 250.00  | cop[tikv] | table:t                  | keep order:false, stats:pseudo                          |
+-------------------------------+---------+-----------+--------------------------+---------------------------------------------------------+
```

The condition `b = 'b'` should be able to pushed down to `IndexScan` phase to reduce the lookup table cost.

We only need to make sure the equal condition value of `b` is less than the index column length 255.

The ideal plan should be

```
mysql> explain select * from t where a between 'a' and 'b' and b = 'b';
+--------------------------+---------+-----------+--------------------------+-------------------------------------------------+
| id                       | estRows | task      | access object            | operator info                                   |
+--------------------------+---------+-----------+--------------------------+-------------------------------------------------+
| IndexReader_7            | 0.25    | root      |                          | index:Selection_6                               |
| └─Selection_6            | 0.25    | cop[tikv] |                          | eq(test.t.b, "b")                               |
|   └─IndexRangeScan_5     | 250.00  | cop[tikv] | table:t, index:a_b(a, b) | range:["a","b"], keep order:false, stats:pseudo |
+--------------------------+---------+-----------+--------------------------+-------------------------------------------------+
```

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by xuyifangreeneyes**
### What problem does this PR solve?

Issue Number: close #21145

Problem Summary:

### What is changed and how it works?

What's Changed:

The conditions like `a = "xxx"` or `b < "yyy"` can be pushed down to IndexScan for the corresponding prefix index if the length of literal string is less than the length of prefix.

How it Works:

Modify the function `splitIndexFilterConditions`, which divides conditions to IndexScan parts and TableScan parts.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: util/ranger/ranger_test.go [hunk 1]
```diff
@@ -2447,3 +2450,236 @@ func TestRangeFallbackForBuildColumnRange(t *testing.T) {
...
+		//	filterConds: "[]",
+		//	resultStr:   "[[\"a\" NULL,\"a\" NULL]]",
+		//},
+		//{
+		//	indexPos:    1,
+		//	exprStr:     "a = 'a' and c is not null",
+		//	accessConds: "[eq(test.t.a, a) not(isnull(test.t.c))]",
+		//	filterConds: "[]",
+		//	resultStr:   "[[\"a\" -inf,\"a\" +inf]]",
+		//},
```
Comment (winoros) [inline util/ranger/ranger_test.go:2589]: Why is commented out?

Comment (xuyifangreeneyes) [reply util/ranger/ranger_test.go:2589]: The test seems broken and I'm fixing it.

Comment (xuyifangreeneyes) [reply util/ranger/ranger_test.go:2589]: Let me update it.

Comment (xuyifangreeneyes) [reply util/ranger/ranger_test.go]: Done.

**Comment by xuyifangreeneyes** [comment]
I close it due to https://github.com/pingcap/tidb/issues/21145#issuecomment-1286696348

**Pull request closed without merging**